### PR TITLE
Update MSYS2 runtime to 2.6.1

### PR DESCRIPTION
**Pull request by dscho**
This Pull Request brings the MSYS2 runtime up-to-date with Cygwin 2.6.1.

While at it, it also addresses two bugs in the POSIX->Windows conversion and backports a fix necessary to adjust for upcoming 64-bit Windows versions.

### CODE REVIEW

**Comment by dscho** [comment]
To make it easier to validate the changes in the patch sets, the first commit adds a script to generate the patch sets from the current `HEAD` in `src/msys2-runtime`.

The 2nd, 3rd and 5th commits used that script to re-generate the patch set from tags [pr-786-step-1](https://github.com/dscho/msys2-runtime/commits/pr-786-step-1), [pr-786-step-3](https://github.com/dscho/msys2-runtime/commits/pr-786-step-2) and [pr-786-step-3](https://github.com/dscho/msys2-runtime/commits/pr-786-step-3) (that I pushed to https://github.com/dscho/msys2-runtime), respectively.

I realize that it is still pretty cumbersome to validate those changes, thanks to the way MSYS2 runtime's patch set is handled these days. If there is anything I can do on my side to make it less cumbersome, please let me know.

**Comment by dscho** [comment]
Seems AppVeyor is not able to build the MSYS2 runtime, even if the very same code works here. I cannot even figure out from AppVeyor's output what the problem is. Can anybody enlighten me?

**Comment by elieux** [comment]
Looks like the actual command that failed was at (https://ci.appveyor.com/project/Alexpux/msys2-packages/build/1.0.659#L2432)[line 2432]:

```
x86_64-w64-mingw32-gcc -s -g -O2 -pipe -ggdb -nostdlib -Wl,-shared -e DllMain -o cyglsa64.dll cyglsa64.o /c/projects/msys2-packages/msys2-runtime/src/msys2-runtime/winsup/lsaauth/cyglsa64.def -ladvapi32 -lkernel32 -lntdll
/opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/bin/ld: /opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libadvapi32.a(ADVAPI32.dll): string not null terminated in ILF object file.
/opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/bin/ld: /opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libadvapi32.a(ADVAPI32.dll): Recognised but unhandled machine type (0x8664) in Import Library Format archive
/opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libadvapi32.a: error adding symbols: File format not recognized
collect2: error: ld returned 1 exit status
[snip]
make[3]: *** [Makefile:75: cyglsa64.dll] Error 1
make[3]: Leaving directory '/c/projects/msys2-packages/msys2-runtime/src/build-x86_64-pc-msys/x86_64-pc-msys/winsup/lsaauth'
make[2]: *** [Makefile:81: lsaauth] Error 1
make[2]: *** Waiting for unfinished jobs....
[big snip]
make[1]: *** [Makefile:9464: all-target-winsup] Error 2
make[1]: Leaving directory '/c/projects/msys2-packages/msys2-runtime/src/build-x86_64-pc-msys'
make: *** [Makefile:883: all] Error 2
```

**Comment by PiotrVV** [comment]
[0002-Rename-DLL-from-cygwin-to-msys.zip](https://github.com/Alexpux/MSYS2-packages/files/704537/0002-Rename-DLL-from-cygwin-to-msys.zip)
small help?

**Comment by elieux** [comment]
I don't understand, @PiotrVV. Are you asking for help, or trying to give help?

**Comment by PiotrVV** [comment]
give help.

**Comment by elieux** [comment]
Yeah. :) I see some patches inside. Can you please explain what they do, or what you changed?

**Comment by dscho** [comment]
> /opt/lib/gcc/x86_64-w64-mingw32/5.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libadvapi32.a: error adding symbols: File format not recognized collect2: error: ld returned 1 exit status

@elieux thank you for digging that out (I was unable to find it in a browser). That error does not occur here, maybe there is something totally going wrong only in AppVeyor?

**Comment by PiotrVV** [comment]
This is error in patch long time ago.

**Comment by mati865** [comment]
@dscho MSYS2 with AppVeyor isn't bug free. It could be random unreproducible error.

**Comment by dscho** [comment]
@PiotrVV Unfortunately, this is not helpful at all. It would have been much easier on everybody involved if you had pushed a proper branch, and cleaned up your changes as well as explained what the heck you are trying to fix.

For the record, this is the diff relative to the tip of my PR (see below for my analysis):

```diff
diff --git a/msys2-runtime/0002-Rename-DLL-from-cygwin-to-msys.patch b/msys2-runtime/0002-Rename-DLL-from-cygwin-to-msys.patch
index bab2ffe2..6bd379c0 100644
--- a/msys2-runtime/0002-Rename-DLL-from-cygwin-to-msys.patch
+++ b/msys2-runtime/0002-Rename-DLL-from-cygwin-to-msys.patch
@@ -1,6 +1,6 @@
-From 5f1c5ffe5310a250b0552b392c3d9d01652a80a7 Mon Sep 17 00:00:00 2001
-From: Johannes Schindelin <johannes.schindelin@gmx.de>
-Date: Fri, 13 Jan 2017 12:40:58 +0100
+From b6e6cb27503d4cae9e7346dc1bbed2e5e9972d35 Mon Sep 17 00:00:00 2001
+From: Alexpux <alexey.pawlow@gmail.com>
+Date: Tue, 5 Apr 2016 10:15:20 +0300
 Subject: [PATCH 02/24] Rename DLL from cygwin to msys
 
 ---
@@ -495,23 +495,31 @@ index 3ec3249..3d2eaf3 100644
  %%
  #undef BRACK
 diff --git a/winsup/cygwin/dlfcn.cc b/winsup/cygwin/dlfcn.cc
-index 159d4fe..7280bab 100644
+index 255a6d5..4a90293 100644
 --- a/winsup/cygwin/dlfcn.cc
 +++ b/winsup/cygwin/dlfcn.cc
-@@ -147,8 +147,13 @@ collect_basenames (pathfinder::basenamelist & basenames,
-   /* If the basename starts with "lib", ... */
-   if (!strncmp (basename, "lib", 3))
-     {
+@@ -144,12 +144,20 @@ collect_basenames (pathfinder::basenamelist & basenames,
+       ext = ".dll";
+       extlen = 4;
+     }
 +#ifdef __MSYS__
+   /* If the basename starts with "lib", ... */
+-  if (!strncmp (basename, "lib", 3))
++  if (!strncmp (basename, "lib", 5))
++  {
 +      /* ... replace "lib" with "msys-", before ... */
-+      basenames.appendv ("msys-", 3, basename+3, baselen-3, ext, extlen, NULL);
++      basenames.appendv ("msys-", 5, basename+5, baselen-5, ext, extlen, NULL);
++  }
 +#else
++	if (!strncmp (basename, "lib", 3))
+     {
        /* ... replace "lib" with "cyg", before ... */
        basenames.appendv ("cyg", 3, basename+3, baselen-3, ext, extlen, NULL);
-+#endif
      }
++#endif
    /* ... using original basename with new suffix. */
    basenames.appendv (basename, baselen, ext, extlen, NULL);
+ }
 diff --git a/winsup/cygwin/dll_init.cc b/winsup/cygwin/dll_init.cc
 index 0fe5714..418e615 100644
 --- a/winsup/cygwin/dll_init.cc
@@ -560,7 +568,7 @@ index f464618..e6f560f 100644
  	  w32len = cygheap->installation_key.Length / sizeof (WCHAR);
  	  if (w32len
 diff --git a/winsup/cygwin/exceptions.cc b/winsup/cygwin/exceptions.cc
-index 743c732..2dd355a 100644
+index 0f5a890..2677ae7 100644
 --- a/winsup/cygwin/exceptions.cc
 +++ b/winsup/cygwin/exceptions.cc
 @@ -500,14 +500,14 @@ try_to_debug (bool waitloop)
@@ -581,7 +589,7 @@ index 743c732..2dd355a 100644
  	  break;
  	}
 diff --git a/winsup/cygwin/fhandler_tty.cc b/winsup/cygwin/fhandler_tty.cc
-index 1505b8c..96a3210 100644
+index aba469f..266927f 100644
 --- a/winsup/cygwin/fhandler_tty.cc
 +++ b/winsup/cygwin/fhandler_tty.cc
 @@ -464,7 +464,11 @@ fhandler_pty_slave::open (int flags, mode_t)
@@ -732,7 +740,7 @@ index 56b4363..6bc948a 100644
  
  #endif /* __CYGWIN_CYGWIN_DLL_H__ */
 diff --git a/winsup/cygwin/include/cygwin/version.h b/winsup/cygwin/include/cygwin/version.h
-index 6ba602e..2745496 100644
+index 820d28b..1bde8c7 100644
 --- a/winsup/cygwin/include/cygwin/version.h
 +++ b/winsup/cygwin/include/cygwin/version.h
 @@ -487,7 +487,11 @@ details. */
@@ -861,7 +869,7 @@ index 7e763e0..36b4d78 100755
    "END_CYGWIN_VERSION_INFO\n\0";
  cygwin_version_info cygwin_version =
 diff --git a/winsup/cygwin/pinfo.cc b/winsup/cygwin/pinfo.cc
-index 9b74442..3e73196 100644
+index 1ce6809..7dbd57e 100644
 --- a/winsup/cygwin/pinfo.cc
 +++ b/winsup/cygwin/pinfo.cc
 @@ -188,7 +188,7 @@ pinfo::maybe_set_exit_code_from_windows ()
@@ -912,7 +920,7 @@ index c250fdc..a230b35 100644
    va_list args;
  
 diff --git a/winsup/cygwin/sec_auth.cc b/winsup/cygwin/sec_auth.cc
-index 50823c6..acd9a95 100644
+index 468d048..e972e4a 100644
 --- a/winsup/cygwin/sec_auth.cc
 +++ b/winsup/cygwin/sec_auth.cc
 @@ -764,7 +764,7 @@ verify_token (HANDLE token, cygsid &usersid, user_groups &groups, bool *pintern)
@@ -956,7 +964,7 @@ index 50823c6..acd9a95 100644
    ts.SourceIdentifier.LowPart = 0x0103;
  
 diff --git a/winsup/cygwin/syscalls.cc b/winsup/cygwin/syscalls.cc
-index 13bb309..41c1a59 100644
+index ba7c743..0ff0b71 100644
 --- a/winsup/cygwin/syscalls.cc
 +++ b/winsup/cygwin/syscalls.cc
 @@ -365,7 +365,7 @@ try_to_bin (path_conv &pc, HANDLE &fh, ACCESS_MASK access, ULONG flags)
@@ -1225,7 +1233,7 @@ index 0ad73fb..63dd4ca 100644
  
  .PHONY: warn_dumper
 diff --git a/winsup/utils/cygcheck.cc b/winsup/utils/cygcheck.cc
-index d1e27b7..050a268 100644
+index f6adf51..50a68ec 100644
 --- a/winsup/utils/cygcheck.cc
 +++ b/winsup/utils/cygcheck.cc
 @@ -68,8 +68,7 @@ static const char *known_env_vars[] = {
@@ -1430,7 +1438,7 @@ index c83b764..42ffbfd 100644
    wcscpy (dllpath, sysdir);
    wcscpy (dllpath + sysdir_len, dll);
 diff --git a/winsup/utils/path.cc b/winsup/utils/path.cc
-index a76cf76..4da4288 100644
+index 4877f6a..dc71dae 100644
 --- a/winsup/utils/path.cc
 +++ b/winsup/utils/path.cc
 @@ -598,14 +598,14 @@ read_mounts ()
@@ -1495,7 +1503,7 @@ index 548b89a..09ee26d 100644
      "The \"-b\" means 'skip the help pages'.  You can omit this until you're\n"
      "familiar with the report layout.  The gprof documentation explains\n"
 diff --git a/winsup/utils/strace.cc b/winsup/utils/strace.cc
-index 4046cce..50ac4a7 100644
+index 5d4a23d..0c8d75e 100644
 --- a/winsup/utils/strace.cc
 +++ b/winsup/utils/strace.cc
 @@ -282,7 +282,7 @@ load_cygwin ()
diff --git a/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch b/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch
index 546d8851..2c8f1dd4 100644
--- a/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch
+++ b/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch
@@ -1,6 +1,6 @@
-From 070d0f27a59494acb4a8823d5e085c4f6d2285b5 Mon Sep 17 00:00:00 2001
-From: Johannes Schindelin <johannes.schindelin@gmx.de>
-Date: Fri, 13 Jan 2017 12:41:51 +0100
+From 61c2a851d4472d1f251954bd8f077202147eccfe Mon Sep 17 00:00:00 2001
+From: Ray Donnelly <mingw.android@gmail.com>
+Date: Sun, 23 Aug 2015 20:47:30 +0100
 Subject: [PATCH 19/24] strace.cc: Don't set MSYS=noglob
 
 Commit message for this code was:
@@ -22,11 +22,11 @@ See quoted () comment:
 "This must have been run from a Windows shell, so preserve
  quotes for globify to play with later."
 ---
- winsup/utils/strace.cc | 20 +++++++++++++++++++-
- 1 file changed, 19 insertions(+), 1 deletion(-)
+ winsup/utils/strace.cc | 19 +++++++++++++++++++
+ 1 file changed, 19 insertions(+)
 
 diff --git a/winsup/utils/strace.cc b/winsup/utils/strace.cc
-index 50ac4a7..484eacb 100644
+index 0c8d75e..f7c98bd 100644
 --- a/winsup/utils/strace.cc
 +++ b/winsup/utils/strace.cc
 @@ -352,11 +352,28 @@ create_child (char **argv)
@@ -50,23 +50,10 @@ index 50ac4a7..484eacb 100644
 +  See quoted () comment:
 +  "This must have been run from a Windows shell, so preserve
 +     quotes for globify to play with later."
- 
++*/
    const char *cygwin_env = getenv ("MSYS");
    const char *space;
  
--  if (cygwin_env && strlen (cygwin_env) <= 256) /* sanity check */
-+  if (cygwin_env && strlen (cygwin_env) <= 256) // sanity check
-     space = " ";
-   else
-     space = cygwin_env = "";
-@@ -364,6 +381,7 @@ create_child (char **argv)
- 				  + strlen (space) + strlen (cygwin_env));
-   sprintf (newenv, "MSYS=noglob%s%s", space, cygwin_env);
-   _putenv (newenv);
-+*/
-   ret = CreateProcess (0, one_line.buf,	/* command line */
- 		       NULL,	/* Security */
- 		       NULL,	/* thread */
 -- 
 2.9.0
 ```

So it seems that @PiotrVV tries to point out that there is an incorrect change in authorship (my rebases somehow changed authorship when I fixed the merge conflicts).

The changes in the file SHA-1s are completely bogus and incorrect.

There seem to be only two real, i.e. functional, changes in the diff:

1)

```diff
 +#ifdef __MSYS__
+   /* If the basename starts with "lib", ... */
+-  if (!strncmp (basename, "lib", 3))
++  if (!strncmp (basename, "lib", 5))
++  {
 +      /* ... replace "lib" with "msys-", before ... */
-+      basenames.appendv ("msys-", 3, basename+3, baselen-3, ext, extlen, NULL);
++      basenames.appendv ("msys-", 5, basename+5, baselen-5, ext, extlen, NULL);
++  }
 +#else
++      if (!strncmp (basename, "lib", 3))
+     {
```

This change is incorrect. It would replace the condition that looks for the prefix "lib" to verify that the first *five* bytes are matching "lib", which is patently wrong. It is no surprise that the change to replace the first *five* characters by "msys-" is wrong, too.

What *is* correct is that my patch should have changed the *first* `3` in the `appendv()` call (and *only* the first one) to `5`.

The second functional change is:

```diff
@@ -50,23 +50,10 @@ index 50ac4a7..484eacb 100644
 +  See quoted () comment:
 +  "This must have been run from a Windows shell, so preserve
 +     quotes for globify to play with later."
- 
++*/
    const char *cygwin_env = getenv ("MSYS");
    const char *space;
  
--  if (cygwin_env && strlen (cygwin_env) <= 256) /* sanity check */
-+  if (cygwin_env && strlen (cygwin_env) <= 256) // sanity check
-     space = " ";
-   else
-     space = cygwin_env = "";
-@@ -364,6 +381,7 @@ create_child (char **argv)
-                                 + strlen (space) + strlen (cygwin_env));
-   sprintf (newenv, "MSYS=noglob%s%s", space, cygwin_env);
-   _putenv (newenv);
-+*/
```

In other words, it tries to *not* comment-out the entire noglob setting. However, this is what the [original patch](https://github.com/Alexpux/Cygwin/commit/61c2a851d4472d1f251954bd8f077202147eccfe) does: it comments out the entire block up to, and including, the `_putenv()` call.

I will fix the attributions as well as the replacement (after digging a bit deeper and verifying that the second parameter to `appendv()` specifies the length of the first parameter) and force-push.

**Comment by dscho** [comment]
> This is error in patch long time ago.

@PiotrVV these non-explanations do not fail to frustrate everybody involved.

**Comment by dscho** [comment]
As promised, I amended the latest commit after fixing the authorship and the parameter to the `appendv()` call.

File: msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch [hunk 1]
```diff
@@ -50,12 +50,17 @@ index 0c8d75e..f7c98bd 100644
...
+-  if (cygwin_env && strlen (cygwin_env) <= 256) /* sanity check */
++  if (cygwin_env && strlen (cygwin_env) <= 256) // sanity check
+     space = " ";
+   else
+     space = cygwin_env = "";
+@@ -364,6 +381,7 @@ create_child (char **argv)
+ 				  + strlen (space) + strlen (cygwin_env));
    sprintf (newenv, "MSYS=noglob%s%s", space, cygwin_env);
    _putenv (newenv);
 +*/
```
Comment (Alexpux) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: Why it here?

Comment (dscho) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: Because it was there before.

Please note that the `+` you see there is the `+` in `0019-*.patch`, it is *not* an indicator of an line added by *this* commit.

There is a space before the `+`, and *that space* indicates that this line has not been modified in the `.patch` file.

I agree that it is difficult to review diffs of diffs, and I really wish that `msys2-runtime` would not have moved to patches, but rather to [merging rebases](https://github.com/git-for-windows/build-extra/blob/746321513a062a3b8b3a06bd588273ac8f8f3abe/shears.sh#L12-L18).

Comment (mati865) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: @dscho there are no spaces before `+` in lines 56-63 (compare it to 50-52).
Also look at the original patch https://github.com/Alexpux/MSYS2-packages/blob/master/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch

Comment (dscho) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: @mati865 This is the original patch adding that end-of-comment: https://github.com/Alexpux/MSYS2-packages/blob/8302b167a80d184a23dc0bd00327d1ee040914b1/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch#L61. And this is the patch after this PR, adding the same end-of-comment: https://github.com/dscho/MSYS2-packages/blob/update-msys2-runtime/msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch#L66

In other words, the hunk before this PR looks like this:

```diff
@@ -360,6 +378,7 @@ create_child (char **argv)
   char *newenv = (char *) malloc (sizeof ("MSYS=noglob") + strlen (space) + strlen (cygwin_env));
   sprintf (newenv, "MSYS=noglob%s%s", space, cygwin_env);
   _putenv (newenv);
+*/
   ret = CreateProcess (0, one_line.buf,	/* command line */
 		       NULL,	/* Security */
 		       NULL,	/* thread */
-- 
2.9.0
```

And afterwards:

```diff
@@ -364,6 +381,7 @@ create_child (char **argv)
 				  + strlen (space) + strlen (cygwin_env));
   sprintf (newenv, "MSYS=noglob%s%s", space, cygwin_env);
   _putenv (newenv);
+*/
   ret = CreateProcess (0, one_line.buf,	/* command line */
 		       NULL,	/* Security */
 		       NULL,	/* thread */
-- 
2.9.0
```

As you will appreciate, both hunks add the end of the comment in exactly the same place: after the `_putenv()` and before the `CreateProcess()`.

**Comment by dscho** [comment]
> Looks like the actual command that failed was at line 2432:

@elieux I actually think that the real culprit comes [much earlier](https://ci.appveyor.com/project/Alexpux/msys2-packages/build/1.0.659#L42):

```
      3 [main] pacman (2156) C:\msys64\usr\bin\pacman.exe: *** fatal error - cygheap base mismatch detected - 0x180339400/0x1802FE408.
```

In other words, Pacman fails to update, and therefore the files in question may not actually be correct (I am still puzzled that the linker complains about an invalid file format, obsolete file versions should still have the correct file format, after all...).

**Comment by elieux** [comment]
> @elieux I actually think that the real culprit comes much earlier:

[snip]

> In other words, Pacman fails to update, and therefore the files in question may not actually be correct (I am still puzzled that the linker complains about an invalid file format, obsolete file versions should still have the correct file format, after all...).

I would argue that this error is not significant. The failure occurs because `bash`&`msys2-runtime` are updated preceding the post-install script for `filesystem`. This script only displayed a few warnings AFAIK. The update process should be otherwise successful.

I agree with @mati865 that it could very well be a random bug. If there was a way to re-run the build on AppVeyor, I'd suggest that (maybe push to the PR branch?).

**Comment by elieux** [comment]
> I agree that it is difficult to review diffs of diffs, and I really wish that msys2-runtime would not have moved to patches, but rather to merging rebases.

I really think that storing patches in the PKGBUILD repository is the better way (compared to just pointing to alexpux/cygwin#msys2-master), but maybe the process can be this: Send and accept (rebase) PRs to alexpux/cygwin and then internally regenerate patches in alexpux/msys2-packages.

**Comment by dscho** [comment]
> I really think that storing patches in the PKGBUILD repository is the better way (compared to just pointing to alexpux/cygwin#msys2-master),

It may be seen as "more correct", but it sure makes it really hard to review, or for that matter, generate PRs.

To counter that "better way", I would like you to understand that Git was written to *avoid* "tarballs and patches", yet that is exactly what you described as better way...

> but maybe the process can be this: Send and accept (rebase) PRs to alexpux/cygwin and then internally regenerate patches in alexpux/msys2-packages.

This makes it even harder, as you simply cannot open a non-conflicting PR after rebasing.

Unless, that is, you followed my link to the description of the "merging rebase", but from your post it does not look like it.

File: msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch (hunk 1 above)
Comment (mati865) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: @dscho, Yes I realized my mistake and edited comment.

Comment (dscho) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: @mati865 Unfortunately I do not get notification mails about edited comments, that's why I missed this edit. Sorry.

Comment (mati865) [inline msys2-runtime/0019-strace.cc-Don-t-set-MSYS-noglob.patch]: No problem.

**Comment by elieux** [comment]
Yes, I didn't have much time to research but wanted to respond nevertheless :/. Let's take this off-topic discussion elsewhere (you choose the communication platform).

@alexpux, is the PR okay to merge?

**Comment by dscho** [comment]
> Yes, I didn't have much time to research but wanted to respond nevertheless :/. Let's take this off-topic discussion elsewhere (you choose the communication platform).

@elieux assuming you were addressing me: I opened a private chat on Gitter.

But let me state for the public record, that I will use whatever scheme you choose to use, and I did not mean to prod you into changing it. I did have to point out the shortcomings of the chosen route, is all.

**Comment by dscho** [comment]
> @Alexpux, is the PR okay to merge?

I would also like to know that. In the negative case, I would further like to know what needs to be fixed. I have more critical ("Bad address") bug fixes coming, so this here PR is a blocker.

**Comment by Alexpux** [comment]
I will look at it on next week. I see that cygwins 2.6.1 release is buggy for many users so maybe wait for 2.6.2 release.

**Comment by Alexpux** [comment]
@dscho , I'm rebasee my patches in Cygwin_Branches repository to 2.6.1. Lets close this pull request and I will add rebased patches from that repo. Then you will create new pull request with changes that not in my patches and clarify why we need it.

**Comment by dscho** [comment]
> Then you will create new pull request with changes that not in my patches and clarify why we need it.

Sure. Just note that this is another downside of the patch-based workflow you chose: it cost me a substantial amount of time to get the PR this far, and now you just throw all the work away, asking me to spend even more time on the problem. It may be nice and convenient for you, but I have to admit that that does not match my experience.

**Comment by Alexpux** [comment]
For msys2-runtime I'm rebase patches directly in cygwin repo and then place patches to msys2-runtime PKGBUILD with git format-patch. This is how I'm do releases for msys2-runtime

**Comment by Alexpux** [comment]
So if you want apply any patches to it maybe need create PR on my Cygwin repo for msys2-master branch

**Comment by dscho** [comment]
> So if you want apply any patches to it maybe need create PR on my Cygwin repo for msys2-master branch

That is unfortunately not good enough: I have done that in the past with patches that were clearly intended to be squashed into incorrect earlier patches, but now we have an ever-growing list of patches that introduce bugs only to fix them in a later patch.

**Comment by dscho** [comment]
> > So if you want apply any patches to it maybe need create PR on my Cygwin repo for msys2-master branch
> 
> That is unfortunately not good enough: I have done that in the past with patches that were clearly intended to be squashed into incorrect earlier patches, but now we have an ever-growing list of patches that introduce bugs only to fix them in a later patch.

I think this conversation demonstrates clearly that the direction I had in mind for the MSYS2 runtime to pick up Git for Windows' patches, and eventually for Git for Windows to replace MSYS2's `mingw-w64-git`, is not viable. The positive side of this is that it saves me tons of time :smile:

**Comment by Alexpux** [comment]
Ok, no problem.

**Comment by Alexpux** [comment]
In any case I don't understand what the problem cherry-picking patches from my Cygwin repo after rebasing.

**Comment by dscho** [comment]
> In any case I don't understand what the problem cherry-picking patches from my Cygwin repo after rebasing.

The problem is that I cannot contribute changes that will result in a new `msys2-runtime` in a timely manner. Not a big problem, though: I'll just keep the fork of Git for Windows running indefinitely.

**Comment by Alexpux** [comment]
I mean that patches need to be rebased when we switch to new release, not between rebuilds of the same runtime
